questionpy/_attempt.py

```diff
@@ -67,61 +60,223 @@ def _merge_uis(
     )
 
 
-class Attempt(BaseAttempt, ABC):
+class AttemptProtocol(Protocol):
+    """Defines the properties and methods an attempt must always contain."""
+
+    @property
+    def cache_control(self) -> CacheControl:
+        pass
+
+    @property
+    def placeholders(self) -> dict[str, str]:
+        pass
+
+    @property
+    def css_files(self) -> list[str]:
+        pass
+
+    @property
+    def files(self) -> dict[str, AttemptFile]:
+        pass
+
+    @property
+    def variant(self) -> int:
+        pass
+
+    @property
+    def formulation(self) -> str:
+        pass
+
+    @property
+    def general_feedback(self) -> str | None:
+        pass
+
+    @property
+    def specific_feedback(self) -> str | None:
+        pass
+
+    @property
+    def right_answer_description(self) -> str | None:
+        pass
+
+    @classmethod
+    @abstractmethod
+    def make_attempt_state(cls, question: "Question", variant: int) -> BaseAttemptState:
+        """Create your attempt state."""
+
+    @abstractmethod
+    def score_response(self, *, try_scoring_with_countback: bool = False, try_giving_hint: bool = False) -> None:
+        pass
+
+
+class AttemptStartedProtocol(AttemptProtocol, Protocol):
+    """In addition to [AttemptProtocol][], defines that a newly started attempt must provide its attempt state.
+
+    The attempt state is only generated at attempt start and immutable afterwards, so it must only be defined on the
+    object returned by [Question.start_attempt][].
+    """
+
+    @abstractmethod
+    def to_plain_attempt_state(self) -> dict[str, JsonValue]:
+        """Return a jsonable representation of this attempt's state."""
+
+
+class AttemptScoredProtocol(AttemptProtocol, Protocol):
+    """In addition to [AttemptProtocol][], defines properties and methods which must be set after scoring."""
+
+    @property
+    def scoring_code(self) -> ScoringCode:
+        pass
+
+    @property
+    def scored_inputs(self) -> Mapping[str, ScoredInputModel]:
```

Comment:
Leider fällt mir jetzt erst auf, dass mir in diesem Attribut bzw. dem `ScoredInputModel` ein Fehler im API-Design unterlaufen ist. `state/specific_feedback/right_answer` und eventuell auch `score_max` passen eigentlich nicht zum `AttemptScoredProtocol`. Die Feedbacktexte sind eigentlich im `AttemptUI` enthalten und sollen vom LMS ohne neues Scoren wieder besorgt werden können.

Ich muss hier nochmal gründlich nachdenken.

Reply:
Entferne am besten `specific_feedback/right_answer/score_max`. Wir können die in einem anderen PR noch (innerhalb von `AttemptUi` einführen). Auch `score_max` zu verschieben, gefällt mir ganz gut, weil wir dann dem LMS die Möglichkeit bieten, die User im laufenden Versuch über die maximale Punktzahl eines einzelnen Eingabefeldes zu informieren.

Stimmt, `state` können wir belassen.

Reply:
https://github.com/questionpy-org/questionpy-server/pull/106/commits/730c13ba718ea2195aadaf3105ddd8a705070c0a